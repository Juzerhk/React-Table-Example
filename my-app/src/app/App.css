.App {
  text-align: center;
  background-color: #F8F8FF;
}

.App-header {
  min-height: 96vh;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

td:hover, td:hover span {
  color: #9A1A4A;
}

td{
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  text-size-adjust: 2px;
}

thead {
  background: #395870;
  color: #fff;
}

.react-bs-table-container{
  margin: 35px;
}

.react-bs-table-sizePerPage-dropdown{
  float: left;
  margin-bottom: 6rem;
}
