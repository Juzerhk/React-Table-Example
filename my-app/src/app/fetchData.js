// Function returns the data fetching from the api
export async function  getRestaurantData(link) {
    
    try {
        const response = await fetch(link);
        const restaurantList = await response.json();
        // console.log('function ', restaurantList);
        return restaurantList;
    } catch(e) {
        console.log('Error fetching data from the server: ', e);
        return null;
    }
}

// Big array, containing more number of objects
export const bigArray = [
    [
      {
        "id": 1,
        "name": "Chunky Cheese",
        "desc": "Say cheese!",
        "phone": "555 0505",
        "address": {
          "country": "USA",
          "city": "Quahog",
          "street": "Spooner St",
          "number": 42
        }
      },
      {
        "id": 2,
        "name": "Pete's Pizza",
        "desc": "Enjoy a tasty slice of life",
        "phone": "555 8384",
        "address": {
          "country": "UK",
          "city": "London",
          "street": "Fake St",
          "number": 66
        }
      }
    ],
    [
      {
        "id": 3,
        "name": "Sally's Seafood",
        "desc": "Yarg maytee. Feast yer gills on our octopus pie",
        "phone": "555 0987",
        "address": {
          "country": "Portugal",
          "city": "Lisbon",
          "street": "Phony Road",
          "number": 123
        }
      },
      {
        "id": 4,
        "name": "Billy's Burgers",
        "desc": "Joy between the buns",
        "phone": "555 5748",
        "address": {
          "country": "Germany",
          "city": "Bielefeld",
          "street": "Phony Road",
          "number": 51
        }
      }
    ],
    [
      {
        "id": 5,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 6,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 7,
        "name": "Keith's Quiche",
        "desc": "Farmer's delight",
        "phone": "555 9999",
        "address": {
          "country": "Germany",
          "city": "Bielefeld",
          "street": "Forged Ave",
          "number": 720
        }
      },
      {
        "id": 8,
        "name": "Katie's Cajun",
        "desc": "Southern fried warmth",
        "phone": "555 6789",
        "address": {
          "country": "USA",
          "city": "New Orleans",
          "street": "Mock blvd",
          "number": 3
        }
      }
    ],
    [
      {
        "id": 9,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 10,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 11,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 12,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 13,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 14,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 15,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 16,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 17,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 18,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 19,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 50
        }
      },
      {
        "id": 20,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 43
        }
      }
    ],
    [
      {
        "id": 21,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 21
        }
      },
      {
        "id": 22,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 22
        }
      }
    ],
    [
      {
        "id": 23,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 23
        }
      },
      {
        "id": 24,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 24
        }
      }
    ],
    [
      {
        "id": 25,
        "name": "Patty's Patties",
        "desc": "Emphatic patties on the patio",
        "phone": "555 4433",
        "address": {
          "country": "Slovakia",
          "city": "Bratislava",
          "street": "Hostel Road",
          "number": 25
        }
      },
      {
        "id": 26,
        "name": "Shelly's Shellfish",
        "desc": "The world's your oyster",
        "phone": "555 2002",
        "address": {
          "country": "Italy",
          "city": "Rome",
          "street": "Via Bogus",
          "number": 26
        }
      }
    ],
  ]