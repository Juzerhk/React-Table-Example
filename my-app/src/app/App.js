import React from 'react';
import './App.css';
import { getRestaurantData } from './fetchData';
import Spinner from 'react-bootstrap/Spinner'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table';

const link = "https://waraclecodetesting.azurewebsites.net/api/restaurants";

class App extends React.Component {
  _isMounted= false;

constructor() {
  super()
  this.state = {
    restaurantData: null,
  }
}

componentDidMount() {
    this._isMounted = true;

  getRestaurantData(link)
    .then(data => { 
        if(this._isMounted) {
            this.setState({
              /* Import bigArray from fetchData.js to display more data in the table and
                  then uncomment Line A and comment Line B
              */
              // restaurantData: this.filterRestaurantData(bigArray),  // Line A
              restaurantData: this.filterRestaurantData(data),  // Line B
            })  
          }
      })
  }
  
  componentWillUnmount(){
    this._isMounted = false;
  }

  flattenArray(array, depth, currentDepth = 0) {
    if (currentDepth === depth) {
      return array;
    }
  
    return array.reduce((acc, item) => {
      return acc.concat(
        Array.isArray(item) ? this.flattenArray(item, depth, currentDepth + 1) : item
      );
    }, []);
  }

  /**
   * Function used to filter the data array as per the criteria
   * All the cities with bielefeld are filtered out
   */
  filterRestaurantData(data) {
    // Uncomment the console statements in this func to see the difference
    // console.log('Before Filtering', data);

    if (data !== null) {
        var filteredData =  this.flattenArray(data).filter((dataFilter) => dataFilter.address.city !== 'Bielefeld');
        /* Wanted to use .flat() of javascript which flattens the array but it throws error
           when running the test case since it only works with latest browser hence not using it 
        */
        // var filteredData = data.flat(Infinity).filter((dataFilter) => dataFilter.address.city !== 'Bielefeld');
        // console.log('After Filtering', filteredData);
        return filteredData;      
    } return [];
  }

  checkDescriptionLength(desc) {
    if(desc.length > 20) {
      desc = desc.slice(0, 20) + '...';
      return desc;
    }
    return desc;
  }

  showDescription(cell, row, name) {
    if(name === 'country'){
      return cell.country;
    } else if (name === 'city'){
      return cell.city;
    } else if (name === 'street'){
      return cell.street;
    } else if (name === 'number'){
      return cell.number;
    }
  }

    /**
   * Function returns the table along with its content
   * Also displays error message if data is not available
   */
  getTable() {
    const { restaurantData } = this.state;

    const options = {
        page: 1,  // which page you want to show as default
        sizePerPageList: [ {
          text: '5', value: 5
        }, {
          text: '10', value: 10
        }, {
          text: 'All', value:  restaurantData.length
        } ], // you can change the dropdown list for size per page
        sizePerPage: 5,  // which size per page you want to locate as default
        pageStartIndex: 1, // where to start counting the pages
        paginationSize: 3,  // the pagination bar size.
        prePage: 'Prev', // Previous page button text
        nextPage: 'Next', // Next page button text
        firstPage: 'First', // First page button text
        lastPage: 'Last', // Last page button text
        paginationShowsTotal: this.renderShowsTotal,  // Accept bool or function
        paginationPosition: 'bottom'  // default is bottom, top and both is all available
        // hideSizePerPage: true > You can hide the dropdown for sizePerPage
        // alwaysShowAllBtns: true // Always show next and previous button
        // withFirstAndLast: false > Hide the going to First and Last page button
      };

    return(
      restaurantData.length > 0 ?
            <BootstrapTable data={restaurantData} pagination={ true } options={ options } >
              <TableHeaderColumn row='0' rowSpan="2" isKey dataField='name'>Name</TableHeaderColumn>
              <TableHeaderColumn row='0' rowSpan="2" dataField='desc' columnTitle>Description</TableHeaderColumn>
              <TableHeaderColumn row='0' rowSpan="2" dataField='phone'>Phone</TableHeaderColumn>
              <TableHeaderColumn row='0' colSpan="4" thStyle={ { textAlign: 'center' } } >Address</TableHeaderColumn>
              <TableHeaderColumn row='1' dataField="address" dataFormat={(cell, row)=>this.showDescription(cell,row,'country')}>Country</TableHeaderColumn>
              <TableHeaderColumn row='1' dataField="address" dataFormat={(cell, row) => this.showDescription(cell,row,'city')}>City</TableHeaderColumn>
              <TableHeaderColumn row='1' dataField="address" dataFormat={(cell, row) => this.showDescription(cell,row,'street')}>Street</TableHeaderColumn>
              <TableHeaderColumn row='1' dataField="address" dataFormat={(cell, row) => this.showDescription(cell,row,'number')}>Number</TableHeaderColumn>
            </BootstrapTable>
            : 
            <div>
              Oops, sorry unable to find any data
            </div>
          );
    }

  render() {
      return (
        <div className="App">
          <header className="App-header">
            <a href="/" >
              <img src={require('./resources/52172.svg')} alt=""></img>
            </a>
            <h2>Waracle's Restaurant List</h2> 
            {
              this.state.restaurantData != null ?
                this.getTable()
                :
                <Spinner animation="grow" />
              }
          </header>
          <footer style={{fontStyle: 'italic'}}>
            Â©Copyright 2019 Juzer Kanchwala All Rights Reserved
          </footer>
        </div>
      );
    }
}
export default App;
