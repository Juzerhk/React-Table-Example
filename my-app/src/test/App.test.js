import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App/App';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import expect from 'expect'; 
import {getRestaurantData} from '../app/fetchData.js';
configure({ adapter: new Adapter() });

const link = "https://waraclecodetesting.azurewebsites.net/api/restaurants";

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("should render initial layout", () => {
  const component = shallow(<App />);
  expect(component.getElements()).toMatchSnapshot();
});

  it('should verify header title', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find('header').text()).toEqual("Waracle's Restaurant List");
  });
  
  it('api call test',  async () => {
    const result = await getRestaurantData(link);
    await expect(result.length).toBeGreaterThan(0);
    await expect(Array.isArray(result)).toBe(true);
  });

