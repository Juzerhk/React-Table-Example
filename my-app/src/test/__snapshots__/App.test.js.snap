// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render initial layout 1`] = `
Array [
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <a
        href="/"
      >
        <img
          alt=""
          src={
            Object {
              "ReactComponent": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "__esModule": true,
              "default": "52172.svg",
            }
          }
        />
      </a>
      <h2>
        Waracle's Restaurant List
      </h2>
      <ForwardRef
        animation="grow"
      />
    </header>
    <footer
      style={
        Object {
          "fontStyle": "italic",
        }
      }
    >
      Â©Copyright 2019 Juzer Kanchwala All Rights Reserved
    </footer>
  </div>,
]
`;
